### PR TITLE
[SE-0060] Enforcing order of defaulted parameters

### PR DESCRIPTION
**Pull request by rintaro**
#### What's in this pull request?

Implement [[SE-0060] Enforcing order of defaulted parameters](https://github.com/apple/swift-evolution/blob/master/proposals/0060-defaulted-parameter-order.md).

Commit 1,2: Ground work: Eliminate existing out of order arguments in Swift source tree.
Commit 3: Disallow out-of-order arguments regardless of whether they have default arguments.
Commit 4: Add `fixItExchange()` facility in `InFlightDiagnostic`
Commit 5: Improve diagnostics for out-of-order arguments. Added fix-it.

With this PR:

``` swift
func fn(_ x: Int, y: Int, _ z: Int...) { }
fn(1, 2, 3, 4, 5, 6, y: 7)
```

emits a diagnostic:

``` swift
test.swift:2:22: error: argument 'y' must precede unnamed parameter #1
fn(1, 2, 3, 4, 5, 6, y: 7)
      ~~~~~~~~~~~~~  ^~~~
      y: 7           2, 3, 4, 5, 6
```
#### Resolved bug number: ([SR-1489](https://bugs.swift.org/browse/SR-1489))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

File: utils/gyb_stdlib_unittest_support.py [hunk 1]
```diff
@@ -8,9 +8,10 @@
 # See http://swift.org/LICENSE.txt for license information
 # See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
 
-TRACE = '''_ message: @autoclosure () -> String = "",
-  showFrame: Bool = true,
+TRACE = '''
+  _ message: @autoclosure () -> String = "",
   stackTrace: SourceLocStack = SourceLocStack(),
+  showFrame: Bool = true,
```
Comment (rintaro) [inline utils/gyb_stdlib_unittest_support.py]: @gribozavr This re-ordering is acceptable?

**Comment by rintaro** [comment]
CC: @jckarter the author of the evolution proposal.

**Comment by jckarter** [comment]
Thanks for taking care of this, @rintaro! Looks good to me, though @DougGregor is the expert in these matter.s

**Comment by jckarter** [comment]
@swift-ci Please test

**Comment by rintaro** [comment]
Oops, I will create a PR on `swift-corelibs-foundation`

https://ci.swift.org/job/swift-PR-Linux/2374/console
```
TestFoundation/TestNSNotificationQueue.swift TestFoundation/TestNSTextCheckingResult.swift TestFoundation/TestNSTask.swift TestFoundation/TestNSRange.swift -o ../buildbot_linux/foundation-linux-x86_64/TestFoundation/TestFoundation
TestFoundation/TestNSAffineTransform.swift:47:36: error: unnamed parameter #5 must precede argument 'line'
                                   "x (expected: \(expectedPoint.x), was: \(newPoint.x)): \(message)")
                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                   line: line
TestFoundation/TestNSAffineTransform.swift:49:36: error: unnamed parameter #5 must precede argument 'line'
                                   "y (expected: \(expectedPoint.y), was: \(newPoint.y)): \(message)")
                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                   line: line
TestFoundation/TestNSAffineTransform.swift:55:36: error: unnamed parameter #5 must precede argument 'line'
                                   "width (expected: \(expectedSize.width), was: \(newSize.width)): \(message)")
                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                   line: line
TestFoundation/TestNSAffineTransform.swift:57:36: error: unnamed parameter #5 must precede argument 'line'
                                   "height (expected: \(expectedSize.height), was: \(newSize.height)): \(message)")
                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                   line: line
TestFoundation/TestNSAffineTransform.swift:64:34: error: unnamed parameter #5 must precede argument 'line'
                                 "origin (expected: \(expectedRect.origin), was: \(newRect.origin)): \(message)")
                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                 line: line
TestFoundation/TestNSAffineTransform.swift:66:33: error: unnamed parameter #5 must precede argument 'line'
                                "size (expected: \(expectedRect.size), was: \(newRect.size)): \(message)")
                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                line: line
```

File: lib/Sema/CSSimplify.cpp [hunk 2]
```diff
@@ -540,19 +540,9 @@ matchCallArguments(ArrayRef<CallArgParam> args,
...
-          continue;
-
-        unsigned prevArgIdx = parameterBindings[prevParamIdx].front();
-        listener.outOfOrderArgument(argIdx, prevArgIdx);
-        return true;
-      }
+      unsigned prevArgIdx = parameterBindings[prevParamIdx].front();
+      listener.outOfOrderArgument(argIdx, prevArgIdx);
+      return true;
     }
```
Comment (DougGregor) [inline lib/Sema/CSSimplify.cpp]: I love deleting code!

**Comment by DougGregor** [comment]
This looks great. Please feel free to merge once the swift-corelibs-foundation change is in.

**Comment by jckarter** [comment]
Thanks @DougGregor. @rintaro Ping me when the corelibs change is committed and I'll spin CI again.

**Comment by rintaro** [comment]
PR for `swift-corelibs-foundation` change: https://github.com/apple/swift-corelibs-foundation/pull/428

Still, I have a similar issue on OS X Platform too:
```
/Users/buildnode/jenkins/workspace/swift-PR-osx/swift/test/1_stdlib/TestAffineTransform.swift:60:33: error: unnamed parameter #5 must precede argument 'line'
                                "size (expected: \(expectedRect.size), was: \(newRect.size)): \(message)")
                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                line: line
```

**Comment by rintaro** [comment]
@swift-ci Please test OS X platform

File: utils/gyb_stdlib_unittest_support.py (hunk 1 above)
Comment (gribozavr) [inline utils/gyb_stdlib_unittest_support.py]: Standard library source and test changes LGTM.

**Comment by gribozavr** [comment]
Standard library source and test changes LGTM.

File: lib/Sema/CSDiag.cpp [hunk 3]
```diff
@@ -4002,24 +4002,48 @@ static bool diagnoseSingleCandidateFailures(CalleeCandidateInfo &CCI,
...
-      diagLoc = tuple->getElementNameLoc(OOOArgIdx);
+        firstRange.Start = tuple->getElementNameLoc(OOOArgIdx);
+    unsigned OOOParamIdx = argBindings[OOOArgIdx];
+    if (paramBindings[OOOParamIdx].size() > 1)
+        firstRange.End = 
+          tuple->getElement(paramBindings[OOOParamIdx].back())->getEndLoc();
+
+    auto secondRange = tuple->getElement(OOOPrevArgIdx)->getSourceRange();
+    if (!second.empty())
+        secondRange.Start = tuple->getElementNameLoc(OOOPrevArgIdx);
```
Comment (rintaro) [inline lib/Sema/CSDiag.cpp]: I will fix the indent. (2 spaces)

**Comment by rintaro** [comment]
@DougGregor @jckarter @gribozavr Thank you for the review!

Since corelibs-foundation change landed, I'll kick CI now.
@swift-ci Please test

**Comment by rintaro** [comment]
I merged this.

@jckarter 
Could you take care of CHANGELOG?
Because I'm not good at writing in English... 😓

**Comment by jckarter** [comment]
Sure thing. Thanks for doing the implementation, @rintaro.

**Pull request merged**